File: discipline_daemon_lib/src/chronic/countdown.rs
use serde::{Deserialize, Serialize};
use crate::x::{Duration, MonotonicInstant};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CountdownStatus {
  Pending,
  Running,
  Finished,
}

impl CountdownStatus {
  pub fn is_pending(self) -> bool {
    matches!(self, CountdownStatus::Pending)
  }

  pub fn is_running(self) -> bool {
    matches!(self, CountdownStatus::Running)
  }

  pub fn is_finished(self) -> bool {
    matches!(self, CountdownStatus::Finished)
  }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Countdown {
  pub from: MonotonicInstant,
  pub duration: Duration,
}

impl Countdown {
  pub fn create(from: MonotonicInstant, duration: Duration) -> Self {
    Self {
      from,
      duration,
    }
  }

  pub fn construct(from: MonotonicInstant, duration: Duration) -> Countdown {
    Self { 
      from, 
      duration,
    }
  }

  pub fn get_from(&self) -> MonotonicInstant {
    self.from
  }

  pub fn get_till(&self) -> MonotonicInstant {
    self.from.plus_or_max(self.duration)
  }

  pub fn get_total_duration(&self) -> Duration {
    self.duration
  }

  pub fn set_total_duration(&mut self, new_value: Duration) {
    self.duration = new_value;
  }

  pub fn get_duration_till_start_or_zero(&self, now: MonotonicInstant) -> Duration {
    now.till_or_zero(self.from)
  }

  pub fn get_duration_since_start_or_zero(&self, now: MonotonicInstant) -> Duration {
    now.since_or_zero(self.from)
  }

  pub fn get_elapsed_time_or_zero(&self, now: MonotonicInstant) -> Duration {
    self.get_duration_since_start_or_zero(now).min(self.duration)
  }

  pub fn get_remaining_time_or_zero(&self, now: MonotonicInstant) -> Duration {
    self.get_total_duration().minus_or_zero(self.get_elapsed_time_or_zero(now))
  }

  pub fn get_time_till_finish_or_zero(&self, now: MonotonicInstant) -> Duration {
    now.till_or_zero(self.get_till())
  }

  pub fn get_status(&self, now: MonotonicInstant) -> CountdownStatus {
    if now.is_eariler_than(self.from) {
      return CountdownStatus::Pending;
    } 
    
    let elapsed_time = self.get_elapsed_time_or_zero(now);
    if elapsed_time.is_shorter_than_or_equal_to(self.duration) {
      return CountdownStatus::Running;
    }
    
    CountdownStatus::Finished
  }

  // todo: delete
  pub fn get_remaining_duration_or_zero(&self, now: MonotonicInstant) -> Duration {
    match self.get_status(now) {
      CountdownStatus::Pending => {
        self.duration
      }
      CountdownStatus::Running => {
        self.duration.minus_or_zero(self.from.till_or_zero(now))
      }
      CountdownStatus::Finished => {
        Duration::zero()
      }
    }
  }

  pub fn is_pending(&self, now: MonotonicInstant) -> bool {
    self.get_status(now).is_pending()
  }

  pub fn is_running(&self, now: MonotonicInstant) -> bool {
    self.get_status(now).is_running()
  }

  pub fn is_finished(&self, now: MonotonicInstant) -> bool {
    self.get_status(now).is_finished()
  }
}

File: discipline_daemon_primitive_ts/chronic/countdown.ts
import { Duration, Instant } from "../x.ts";

export const enum CountdownStatus {
  Pending,
  Running,
  Finished,
}

const status = Symbol();

type CountdownPending = Countdown & { 
  [status]: CountdownStatus.Pending 
};
type CountdownRunning = Countdown & { 
  [status]: CountdownStatus.Running 
};
type CountdownFinished = Countdown & { 
  [status]: CountdownStatus.Finished 
};

export class Countdown {
  declare [status]: CountdownStatus;

  private constructor(
    private from: Instant,
    private duration: Duration,
  ) {}

  static construct(from: Instant, duration: Duration): Countdown {
    // if (from.isLaterThan(till)) {
    //   return TextualError
    //     .create("Constructing 'Countdown' from 'from' and 'till' arguments")
    //     .addStringAttachment("'from' argument", from.toElapsedTime().toString2())
    //     .addStringAttachment("'till' argument", till.toElapsedTime().toString2());
    // }
    // if (from.isAt(till)) {
    //   return new Error("Creating Countdown from 'from' and 'till' arguments: 'from' is equal to 'till'");
    // }
    return new Countdown(from, duration);
  }

  static create(from: Instant, duration: Duration) {
    return this.construct(from, duration);
  }

  getFrom() {
    return this.from;
  }
  
  getTill() {
    return this.from.plusOrMax(this.duration);
  }

  /**
   * Returns the total duration this Countdown runs for.
   */
  getTotalDuration(): Duration {
    return this.duration;
  }

  setTotalDuration(duration: Duration)  {
    this.duration = duration;
  }
  /**
   * Returns zero if not pending.
   */
  getTimeTillStartOrZero(now: Instant): Duration {
    return now.tillOrZero(this.from);
  }
  /**
   * Returns zero if pending.
   */
  getTimeSinceStartOrZero(now: Instant): Duration {
    return now.sinceOrZero(this.from);
  }
  /**
   * Returns zero if pending.
   * Returns `this.getTotalDuration()` if finished. 
   */
  getElapsedTimeOrZero(now: Instant): Duration {
    return this.getTimeSinceStartOrZero(now).min(this.duration)
  }
  /**
   * if pending, returns `Duration.zero()`.
   * if running, returns how much more time the countdown will run for.
   * if finished, returns `Duration.zero().
   */
  getRemainingTimeOrZero(now: Instant): Duration {
    return this
      .getTotalDuration()
      .minusOrZero(this.getElapsedTimeOrZero(now))
  }
  /**
   * if pending, returns the time from @param now till `this.getTill()`.
   * if running, returns the time from @param now till `this.getTill()`.
   * if finished, returns `Duration.zero()`.
   */
  getTimeTillFinishOrZero(now: Instant) {
    return now.tillOrZero(this.getTill());
  }

  getStatus(now: Instant) {
    if (now.isEarilerThan(this.from)) {
      return CountdownStatus.Pending;
    }

    const elapsedTime = this.from.tillOrZero(now);
    if (elapsedTime.isShorterThanOrEqualTo(this.duration)) {
      return CountdownStatus.Running;
    }

    return CountdownStatus.Finished;
  }

  isPending(now: Instant): boolean {
    return this.getStatus(now) === CountdownStatus.Pending;
  }

  isRunning(now: Instant): boolean {
    return this.getStatus(now) === CountdownStatus.Running;
  }

  isFinished(now: Instant): boolean {
    return this.getStatus(now) === CountdownStatus.Finished;
  }

  extendByOrSetToMax(factor: Duration) {
    this.duration = this.duration.plusOrMax(factor); 
  }

  match<A, B, C>(now: Instant, cases: {
    Pending: (it: CountdownPending) => A,
    Running: (it: CountdownRunning) => B,
    Finished: (it: CountdownFinished) => C,
  }): A | B | C {
    switch (this.getStatus(now)) {
      case CountdownStatus.Pending: return cases.Pending(this as CountdownPending);
      case CountdownStatus.Running: return cases.Running(this as CountdownRunning);
      case CountdownStatus.Finished: return cases.Finished(this as CountdownFinished);
    }
  }
}

